docs/api-docs.rst

```diff
@@ -1,6 +1,217 @@
 API documentation
 =================
 
+Overview
+--------
+
+The main purpose of Romanesco is to execute a broad range of tasks. These tasks,
+along with a set of input bindings and output bindings are passed to the :py:func:`romanesco.run`
+function, which is responsible for fetching the inputs as necessary and executing
+the task, and finally populating any output variables and sending them to their
+destination.
+
+The task, its inputs, and its outputs are each passed into the function as python dictionaries.
+In this section, we describe the structure of each of those dictionaries.
+
+The task specification
+**********************
+
+The first argument to :py:func:`romanesco.run` describes the task to execute,
+independently of the actual data that will it will be executed against. The most
+important field of the task is the ``mode``, which describes what type of task
+this is. The structure for the task dictionary is described below. Uppercase names
+within angle braces represent symbols defined in the specification. Optional parts
+of the specification are surrounded by parenthesis to avoid ambiguity with the
+square braces, which represent lists in python or Arrays in JSON.
+
+.. code-block :: none
+
+    <TASK> ::= <PYTHON_TASK> | <R_TASK> | <DOCKER_TASK> | <WORKFLOW_TASK>
+
+    <PYTHON_TASK> ::= {
+        "mode": "python",
+        "script": <python code to run as a string>
+        (, "inputs": [<TASK_INPUT> (, <TASK_INPUT>, ...)])
+        (, "outputs": [<TASK_OUTPUT> (, <TASK_OUTPUT>, ...)])
```

Comment:
This would be a lot cleaner if JSON allowed for trailing commas, but I like the precision here that following this exactly will result in a spec in proper JSON form.